src/mlpack/core/optimizers/parallel_sgd/parallel_sgd_impl.hpp

```diff
@@ -0,0 +1,148 @@
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_PARALLEL_SGD_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_PARALLEL_SGD_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "parallel_sgd.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+template <typename DecayPolicyType>
+ParallelSGD<DecayPolicyType>::ParallelSGD(
+    const size_t maxIterations,
+    const size_t threadShareSize,
+    const double tolerance,
+    const DecayPolicyType& decayPolicy) :
+    maxIterations(maxIterations),
+    threadShareSize(threadShareSize),
+    tolerance(tolerance),
+    decayPolicy(decayPolicy)
+{ /* Nothing to do. */ }
+
+template <typename DecayPolicyType>
+template <typename SparseFunctionType>
+double ParallelSGD<DecayPolicyType>::Optimize(
+    SparseFunctionType& function,
+    arma::mat& iterate)
+{
+  double overallObjective = DBL_MAX;
+  double lastObjective;
+
+  for (size_t i = 1; i != maxIterations; ++i)
+  {
+    // Calculate the overall objective.
+    lastObjective = overallObjective;
+    overallObjective = 0;
+
+    for (size_t j = 0; j < function.NumFunctions(); ++j)
+    {
+      overallObjective += function.Evaluate(iterate, j);
+    }
+
+    // Output current objective function.
+    Log::Info << "Parallel SGD: iteration " << i << ", objective "
+      << overallObjective << "." << std::endl;
+
+    if (std::isnan(overallObjective) || std::isinf(overallObjective))
+    {
+      Log::Warn << "Parallel SGD: converged to " << overallObjective
+        << "; terminating" << " with failure.  Try a smaller step size?"
+        << std::endl;
+      return overallObjective;
+    }
+
+    if (std::abs(lastObjective - overallObjective) < tolerance)
+    {
+      Log::Info << "SGD: minimized within tolerance " << tolerance << "; "
+        << "terminating optimization." << std::endl;
+      return overallObjective;
+    }
+
+    // Get the stepsize for this iteration
+    double stepSize = decayPolicy.StepSize(i);
+
+    arma::Col<size_t> visitationOrder;
+    GenerateVisitationOrder(visitationOrder, function.NumFunctions());
+
+    #pragma omp parallel
+    {
+      // Each processor gets a subset of the instances.
+      // Each subset is of size batchSize.
+      arma::Col<size_t> instances = ThreadShare(omp_get_thread_num(),
+          visitationOrder);
+      for (size_t j = 0; j < instances.n_elem; ++j)
+      {
+        // Each instance affects only some components of the decision variable.
+        // So the gradient is sparse.
+        arma::sp_mat gradient;
+
+        // Evaluate the sparse gradient.
+        function.Gradient(iterate, instances[j], gradient);
+
+        // Update the decision variable with non-zero components of the
+        // gradient.
+        for (size_t i = 0; i < gradient.n_cols; ++i)
+        {
+          // Iterate over the non-zero elements.
+          for (auto cur = gradient.begin_col(i); cur != gradient.end_col(i);
```

Comment:
In general we should avoid ``auto``when using armadillo objects; from the armadillo page:

> Can I use the C++11 auto keyword with Armadillo objects and/or expressions? 
> Use of C++11 auto is not recommended with Armadillo objects and expressions. Armadillo has a template meta-programming framework which creates lots of short lived temporaries that are not handled by auto.
> 

I think in this case it's okay, but it would be easy to use the type of the iterator instead of ``auto``. Let me know what you think.